src/wallet/coinselection.cpp

```diff
@@ -102,12 +102,19 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
     bool is_feerate_high = utxo_pool.at(0).fee > utxo_pool.at(0).long_term_fee;
     bool max_tx_weight_exceeded = false;
 
+    // We seek for a solution where the process does not create change, so we can go as high as one unit below
+    // the amount where the transaction creation process starts creating a change output.
+    // In other words, the range of the possible solutions is:
+    // target <= solution_amount < target + min_viable_change - 1.
+    // Note: when 'min_viable_change=0', the upper bound is 'selection_target'.
+    CAmount selection_upper_bound = selection_target + (min_viable_change != 0 ? min_viable_change - 1 : 0);
```

Comment:
In `GetChange`, `min_viable_change` refers to the minimum amount of the change output (after deducting the fee), so I think this here should at least be `selection_target + min_viable_change + change_fee`. And `min_viable_change` should be at least `change_spend_fee + 1`.

Since `cost_of_change` is `change_fee  + changes_spend_fee`, that would make `min_viable_change` exactly 1 bigger than `cost_of_change`.

Generally, I think we should be careful to distinguish the `min_viable_change` from the “budget necessary to create a change output with at least an amount of `min_viable_change`. 

I think this PR may also overlap with https://github.com/bitcoin/bitcoin/pull/28366

Reply:
> In `GetChange`, `min_viable_change` refers to the minimum amount of the change output (after deducting the fee), so I think this here should at least be `selection_target + min_viable_change + change_fee`. And `min_viable_change` should be at least `change_spend_fee + 1`.
> 
> Since `cost_of_change` is `change_fee + changes_spend_fee`, that would make `min_viable_change` exactly 1 bigger than `cost_of_change`.

Great comment. This made me go further over https://github.com/bitcoin/bitcoin/pull/28395#pullrequestreview-1618045956. It seems that we had different views of the process. You were watching the typical flow, while I was looking at the SFFO one.

Will work on it. Thanks!.